### PR TITLE
Terminal Receive pane scrolling improvements and buffer increase

### LINKED ISSUE #199: Increase Receive Buffer Size
After fixing the Receive Pane processing, via issue #178, now we need to increase the internal buffer size to hold more lines so that the user has more time to scroll back in history on a fast-updating display without it overwriting data as soon.

Please make the receive buffer able to hold 10,000 lines before overwriting old lines with new data.

[Here's a video](https://drive.google.com/file/d/0B0tev9Y3WaL4MEhuRGNPeUM0R0E/view?usp=sharing) that illustrates some of the problem experienced apparently due to a shorter receive buffer.

### LINKED ISSUE #202: Debug Terminal: Receive Pane Scrolls Awkwardly
The Receive Pane scrolls awkwardly when there's many lines of text to scroll back to in history.  This is best demonstrated by [this video example](https://drive.google.com/file/d/0B0tev9Y3WaL4QnlyQW9rdVJsb1E/view?usp=sharing) and another video that shows a problem at [the opposite end of the history](https://drive.google.com/file/d/0B0tev9Y3WaL4ZDk2X1d5cVY2RVU/view?usp=sharing).

Note: In addition to the strange looking behavior where the first text appears at the bottom of the display (which I'm not calling a _bug_ by itself) the other side effect inherit in the solution employed causes the scroll thumb to stay staticly sized (and pushed to the bottom of the scroll bar).  Without the scroll bar thumb resizing, there's no quick-glance visual indication of the size of the Receive Pane's history.

I think Blaine mentioned that avoiding the Chrome scroll bar was problematic, but I see solutions that seem to work great, like [jScrollPane](http://jscrollpane.kelvinluck.com/basic.html), [jQuery slimScroll](http://rocha.la/jQuery-slimScroll), [Control.ScrollBar](http://livepipe.net/control/scrollbar), and the one that may be the most most promising, [Tiny Scrollbar](http://baijs.com/tinyscrollbar/) where the size and thumb can be set manually (maybe others do that also).  All of these examples even catch non-focused mouse over + scroll wheel activity and non-focused mouse over + shift + scroll wheel (which is a big plus for easy vertical and horizontal scrolling, respectively).

Similar to what I've said before, I suggest that the Receive Pane display be a logical _view port_ into the internal line buffer- this view port's content is delivered to it dynamically (the Receive Pane never actually has more content than what is visible; unless a little extra helps with scrolling in some way) and the custom scroll bar's thumb size be adjusted by code according to the size of the internal line buffer, and of course the thumb position is synchronized with the view and the view is synchronized with the thumb.  Also, the reclaimed buffer area (for head-meets-tail moments) be limited to a single-line at a time, to improve the visual experience for the user so that the display is not so jumpy when the circular buffer is full.  This assumes the standard focus, keyboard, mouse controls are active for this scrolling region.  I can elaborate more if needed.

### PR DESCRIPTION
**Pull request by 2fast2fourier**
#### What's this PR do?

This PR improves terminal scroll-back to eliminate the performance hit from scrolling backwards in the console. This also increases the buffer size. Additionally, it prevents one form of 'text slipping' when the underlying buffer trims output by maintaining a line offset, now the console text will remain in place as long as the visible text does not fall outside the buffer storage.
#### What are the important parts of the code?

The primary scrolling changes are in the scrolling management class in `lib/scroller.js`. There were minor changes to the console buffer store, increasing buffer size and tracking trim offsets.
#### How should this be tested by the reviewer?

Try scrolling up and down to see how the console reacts. Here's a simple program to generate constant output:

```
'{$STAMP BS2}
'{$PBASIC 2.5}

  PAUSE 1000
  ix VAR WORD
  ix = 0

  DO
    DEBUG "Loop: ", DEC ix, CR
    ix = ix+1
    PAUSE 5
  LOOP
```
#### What are the relevant tickets?

Closes #202 
Closes #199

### CODE REVIEW

File: src/lib/scroller.js [hunk 1]
```diff
@@ -40,32 +17,63 @@ function Scroller(consoleElement) {
...
+};
+
+Scroller.prototype.setLines = function(newLines, offset) {
   this.lines = newLines;
+  this.lineOffset = offset;
   if(this.sticky){
-    this.startPosition = Math.max(0, len - this.minVisible);
-  }else if(len === 1 && newLines[0].length === 0){
+    this.startPosition = Math.max(this.lineOffset, this.lineCount() - this.visibleCount);
+    this.endPosition = this.lineCount();
```
Comment (phated) [inline src/lib/scroller.js:49]: define this above `this.startPosition` so you don't have to call `this.lineCount()` twice and then use `this.endPosition` in the `this.startPosition` calculation.

File: src/lib/scroller.js [hunk 2]
```diff
@@ -40,32 +17,63 @@ function Scroller(consoleElement) {
...
 };
 
+Scroller.prototype.lineCount = function(){
+  return this.lines.length + this.lineOffset;
+};
+
 Scroller.prototype.reset = function(){
-  this.startPosition = Math.max(0, this.lines.length - this.minVisible);
+  this.startPosition = Math.max(0, this.lineCount() - this.visibleCount);
+  this.endPosition = Math.max(0, this.lineCount());
```
Comment (phated) [inline src/lib/scroller.js:75]: define this above `this.startPosition` so you don't have to call `this.lineCount()` twice and then use `this.endPosition` in the `this.startPosition` calculation.

File: src/lib/scroller.js [hunk 3]
```diff
@@ -80,34 +88,61 @@ Scroller.prototype.requestRefresh = function(){
 Scroller.prototype._renderVisible = function(){
   this.animateRequest = null;
   if(this.dirty && this.console){
-    var top = this.console.scrollTop;
+    const top = this.console.scrollTop;
     if(this.sticky){
-      this.startPosition = Math.max(0, this.lines.length - this.minVisible);
+      this.startPosition = Math.max(this.lineOffset, this.lineCount() - this.visibleCount);
+      this.endPosition = this.lineCount();
```
Comment (phated) [inline src/lib/scroller.js:94]: define this above `this.startPosition` so you don't have to call `this.lineCount()` twice and then use `this.endPosition` in the `this.startPosition` calculation.

File: src/lib/scroller.js [hunk 4]
```diff
@@ -80,34 +88,61 @@ Scroller.prototype.requestRefresh = function(){
...
     }
-    this.console.innerHTML = generateContent(this.lines, this.startPosition, this.lines.length, this.minVisible);
+    this.console.innerHTML = this._generateContent();
     if(this.jumpToBottom){
-      this.console.scrollTop = 2000;
+      this.console.scrollTop = 4000;
```
Comment (phated) [inline src/lib/scroller.js:99]: why was this changed from 2000 to 4000?

File: src/lib/scroller.js [hunk 5]
```diff
@@ -80,34 +88,61 @@ Scroller.prototype.requestRefresh = function(){
...
+  this.endPosition = Math.min(this.lineCount(), this.endPosition + this.visibleCount);
+  if(this.console){
+    // add the new content to the bottom, then get scroll position to remove content
+    this.console.innerHTML = this._generateContent();
+    const scrollHeight = this.console.scrollHeight;
+    const scrollTop = this.console.scrollTop;
+
+    // update start position and render
+    this.startPosition = Math.max(this.lineOffset, Math.min(this.lineCount() - (this.visibleCount * 2), this.endPosition - (this.visibleCount * 2)));
+    this.console.innerHTML = this._generateContent();
```
Comment (phated) [inline src/lib/scroller.js:141]: why is innerHTML being set twice in this method?

File: src/lib/terminal.js [hunk 6]
```diff
@@ -68,7 +70,7 @@ class Terminal {
 
   setCursorPosition(line, col){
     const { lines } = this.state;
-    for(var ix = lines.length; ix <= line; ix++){
+    for(let ix = lines.length; ix <= line; ix++){
```
Comment (phated) [inline src/lib/terminal.js:73]: use `_.fill` here too

File: src/lib/scroller.js (hunk 4 above)
Comment (2fast2fourier) [inline src/lib/scroller.js:99]: The number of visible lines was increased so it wouldn't need as many expansions during a scroll, but if the font was really large it wouldn't scroll properly. It's really unlikely but I just wanted to be safe.

Comment (phated) [inline src/lib/scroller.js:99]: do you see benefit in making this an option attached to the instance?  Something we could configure in the future?

File: src/lib/scroller.js (hunk 5 above)
Comment (2fast2fourier) [inline src/lib/scroller.js:141]: It's a two part update, where it adds new text at the bottom, then grabs the new scrollTop/scrollHeight (which forces the layout). After there it removes the lines above, but that causes the text position to jump. It then uses those scroll values to figure out the exact distance it needs to compensate.

I can't think of a way to do it without forcing two inline render calls, but it's pretty effective at avoiding scroll drift. The performance hit is pretty low too, since it's never rendering more than  50-100 lines at a time.

File: src/lib/terminal.js (hunk 6 above)
Comment (2fast2fourier) [inline src/lib/terminal.js:73]: Fill won't work in this case, fill doesn't work past the bounds of the existing array and this loop is for filling in past the end of the array.

Comment (phated) [inline src/lib/terminal.js:73]: makes sense

File: src/lib/scroller.js (hunk 5 above)
Comment (phated) [inline src/lib/scroller.js:141]: ok

**Comment by phated** [comment]
Everything looks good.  Just want to test on the chromebook to see CPU utilization with the changes but @urbantumbleweed has it currently.  Will test tomorrow before merge.
